uaclient/security_status.py

```diff
@@ -132,6 +149,70 @@ def get_ua_info(cfg: UAConfig) -> Dict[str, Any]:
     return ua_info
 
 
+def get_reboot_status():
+    if not should_reboot():
+        return RebootStatus.REBOOT_NOT_REQUIRED
+
+    try:
+        pkg_list = load_file(REBOOT_PKGS_FILE_PATH)
+    except FileNotFoundError:
+        # If we cannot load that file, we cannot evaluate if we have
+        # kernel related packages that require a reboot. Therefore, we
+        # will just say that a reboot is required
+        return RebootStatus.REBOOT_REQUIRED
+
+    pkg_list = pkg_list.split()
+    num_pkgs = len(pkg_list)
+
+    kernel_regex = "^(linux-image|linux-base).*"
+    num_kernel_pkgs = 0
+
+    for pkg in pkg_list:
+        if re.match(kernel_regex, pkg):
+            num_kernel_pkgs += 1
+
+    # We will only check the Livepatch state if all the
+    # packages that require a reboot are kernel related
+    if num_kernel_pkgs != num_pkgs:
+        return RebootStatus.REBOOT_REQUIRED
+
+    if not which("canonical-livepatch"):
+        return RebootStatus.REBOOT_REQUIRED
+
+    livepatch_status, _ = subp(
+        ["canonical-livepatch", "status", "--format", "json"]
+    )
+
+    try:
+        livepatch_status_dict = json.loads(livepatch_status)
+    except json.JSONDecodeError:
+        # If we cannot parse the json file, we will return a
+        # normal reboot state
+        return RebootStatus.REBOOT_REQUIRED
+
+    patch_statuses = livepatch_status_dict.get("Status", [{}])
+
+    try:
+        kernel_info = get_kernel_info()
+        kernel_name = kernel_info.proc_version_signature_version
+    except UserFacingError:
+        # If we cannot parse the kernel info, we cannot verify
+        # the correct Livepatch status to report
+        kernel_name = ""
+
+    patch_state = ""
+    for patch_status in patch_statuses:
+        if patch_status.get("Kernel") == kernel_name:
+            patch_state = patch_status.get("Livepatch", {}).get("State", "")
+
+    if patch_state == "applied":
+        return RebootStatus.REBOOT_REQUIRED_LIVEPATCH_APPLIED
+
+    # Any other Livepatch status will not be considered here to modify the
+    # reboot state
+    return RebootStatus.REBOOT_REQUIRED
```

Comment:
If the `State` is not `applied` e.g. like on my machine I have `nothing-to-apply` why do we still have to reboot? Or is it that the fact that there are packages on the `REBOOT_PKGS_FILE_PATH` already means we need to reboot?

Reply:
Yes, is the fact that we have the `/var/run/reboot-required` files that state that we need a reboot even if livepatch did nothing on the system